packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -430,9 +444,30 @@ export class FluidDataStoreRuntime
 		this.identifyLocalChangeInSummarizer("DDSCreatedInSummarizer", id, type);
 	}
 
-	public createChannel(id: string = uuid(), type: string): IChannel {
-		if (id.includes("/")) {
-			throw new UsageError(`Id cannot contain slashes: ${id}`);
+	public createChannel(idArg: string | undefined, type: string): IChannel {
+		let id: string;
+
+		if (idArg !== undefined) {
+			id = idArg;
+			this.validateChannelId(id);
+		} else {
+			// We use three non-overlapping namespaces:
+			// - detached state: even numbers
+			// - attached state: odd numbers
+			// - uuids
+			// In first two cases we will encode result as strings in more compact form, with leading underscore,
+			// to ensure no overlap with user-provided DDS names (see validateChannelId())
+			if (this.visibilityState !== VisibilityState.GloballyVisible) {
+				// container is detached, only one client observes content, no way to hit collisions with other clients.
+				id = encodeCompactIdToString(2 * this.contexts.size, "_");
+			} else {
+				// Due to back-compat, we could not depend yet on generateDocumentUniqueId() being there.
+				// We can remove the need to leverage uuid() as fall-back in couple releases.
+				const res =
+					this.dataStoreContext.containerRuntime.generateDocumentUniqueId?.() ?? uuid();
```

Comment:
I think you should add `generateDocumentUniqueId` to `IFluidDataStoreContext` instead to accessing containerRuntime. `containerRuntime` is not accessed anywhere else in this file and we should remove it instead of adding a dependency on it.

Reply:
To be honest, I do not like to re-export such things. While this implementation does not use containerRuntime, this layer (various DataObject implementations) will reach out to Container Runtime for various needs (including ability to create data stores). I think it's fine to have global things on container runtime (though it's possible the example with creation of data stores will not hold water too long :)).
The way how I think about it - context should have everything that parent needs to provide / be ablet to alternate / adapt for a child node. But things that are pretty much never going to change and are the same - it's better to provide through some global accessor. Parent object can still provide alternative implementation, it's just a bit harder.
Let's chat about it more.

Reply:
The fact that data store runtime has to expose container runtime for usage by DataObject, etc. makes this comment invalid. We cannot remote containerRuntime so using it is fine.